timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimeLockServiceMetadataIntegrationTest.java

```diff
@@ -0,0 +1,281 @@
+        IdentifiedLockRequest requestWithoutMetadata = ImmutableIdentifiedLockRequest.copyOf(
+                        standardRequestWithMetadata(ALL_WATCHED_LOCKS_WITH_METADATA))
+                .withMetadata(Optional.empty());
+        waitForFuture(timeLockService.lock(requestWithoutMetadata));
+
+        assertThat(getAllLockEventsMetadata()).containsExactly(Optional.empty());
+    }
+
+    @Test
+    public void canMixWithOtherLockWatchEventsAndAbsentMetadata() {
+        List<Optional<LockRequestMetadata>> metadataList = new ArrayList<>();
+        ALL_WATCHED_LOCKS_WITH_METADATA.forEach((lock, metadata) -> {
+            // -> LockWatchCreatedEvent
+            timeLockService.startWatching(LockWatchRequest.of(
+                    ImmutableSet.of(LockWatchReferences.entireTable("randomTable" + metadataList.size()))));
+
+            // -> LockEvent with metadata
+            Map<LockDescriptor, ChangeMetadata> map = ImmutableMap.of(lock, metadata);
+            IdentifiedLockRequest requestWithMetadata = standardRequestWithMetadata(map);
+            LockResponseV2 response = waitForFuture(timeLockService.lock(requestWithMetadata));
+            metadataList.add(Optional.of(LockRequestMetadata.of(map)));
+
+            // -> UnlockEvent
+            waitForFuture(timeLockService.unlock(ImmutableSet.of(getToken(response))));
+
+            // -> LockEvent with absent metadata
+            IdentifiedLockRequest requestWithoutMetadata = ImmutableIdentifiedLockRequest.copyOf(
+                            standardRequestWithMetadata(map))
+                    .withMetadata(Optional.empty());
+            response = waitForFuture(timeLockService.lock(requestWithoutMetadata));
+            metadataList.add(Optional.empty());
+
+            // -> UnlockEvent
+            waitForFuture(timeLockService.unlock(ImmutableSet.of(getToken(response))));
+        });
+
+        assertThat(getAllLockWatchEvents()).hasSize(5 * ALL_WATCHED_LOCKS_WITH_METADATA.size());
+        assertThat(getAllLockEventsMetadata()).containsExactlyElementsOf(metadataList);
+    }
+
+    private List<Optional<LockRequestMetadata>> getAllLockEventsMetadata() {
+        return getAllLockWatchEvents().stream()
+                .map(event -> event.accept(LOCK_WATCH_EVENT_VISITOR))
+                .flatMap(Optional::stream)
+                .map(LockEvent::metadata)
+                .collect(Collectors.toList());
+    }
+
+    private List<LockWatchEvent> getAllLockWatchEvents() {
+        ConjureStartTransactionsResponse response =
+                waitForFuture(timeLockService.startTransactionsWithWatches(startTransactionsRequestWithInitialVersion));
+
+        return response.getLockWatchUpdate().accept(LOCK_WATCH_STATE_UPDATE_VISITOR);
+    }
+
+    private static <T> T waitForFuture(Future<T> future) {
+        Awaitility.await().atMost(Duration.ofMillis(200)).until(future::isDone);
```

Comment:
Yeah let's use direct executors so we don't have a fixed overhead of sleeping!

Reply:
See comment above. I was under the impression that Awaitility polls at the beginning of the sleep, but it apparently doesn't. Changed to assert that the future is done (and return the value).